File: dlp/cpp/prac5_LL1.cpp
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

// Define a grammar using a map where non-terminals map to their productions
unordered_map<char, vector<string>> grammar = {
    {'E', {"aX"}},
    {'X', {"+aX", "c"}}
};

unordered_map<char, unordered_set<char>> FIRST;
unordered_map<char, unordered_set<char>> FOLLOW;

void compute_FIRST() {
    bool changed = true;
    
    while (changed) {
        changed = false;
        
        for (const auto& rule : grammar) {
            char nt = rule.first;
            FIRST[nt].clear();
            
            for (const string& prod : rule.second) {
                char firstChar = prod[0];
                
                if (isupper(firstChar)) {
                    if (FIRST.find(firstChar) != FIRST.end()) {
                        unordered_set<char>& firstSet = FIRST[firstChar];
                        size_t originalSize = firstSet.size();
                        
                        FIRST[nt].insert(firstSet.begin(), firstSet.end());
                        
                        if (firstSet.find('') == firstSet.end()) {
                            break;
                        }
                        
                        if (originalSize != firstSet.size()) {
                            changed = true;
                        }
                    }
                } else {
                    FIRST[nt].insert(firstChar);
                    break;
                }
            }
        }
    }
}

void compute_FOLLOW() {
    bool changed = true;
    
    while (changed) {
        changed = false;
        
        for (const auto& rule : grammar) {
            char nt = rule.first;
            
            for (const string& prod : rule.second) {
                for (size_t i = 0; i < prod.size(); ++i) {
                    char current = prod[i];
                    
                    if (isupper(current)) {
                        if (i == prod.size() - 1) {
                            FOLLOW[current].insert(FOLLOW[nt].begin(), FOLLOW[nt].end());
                        } else {
                            for (size_t j = i + 1; j < prod.size(); ++j) {
                                if (isupper(prod[j])) {
                                    FOLLOW[current].insert(FIRST[prod[j]].begin(), FIRST[prod[j]].end());
                                } else {
                                    FOLLOW[current].insert(prod[j]);
                                    break;
                                }
                                
                                if (FIRST[prod[j]].find('') == FIRST[prod[j]].end()) {
                                    break;
                                }
                            }
                        }
                        
                        if (FOLLOW[current].size() != FOLLOW[current].size()) {
                            changed = true;
                        }
                    }
                }
            }
        }
    }
}

bool is_LL1() {
    compute_FIRST();
    compute_FOLLOW();
    
    for (const auto& rule : grammar) {
        char nt = rule.first;
        
        for (const string& prod1 : rule.second) {
            for (const string& prod2 : rule.second) {
                if (&prod1 != &prod2) {
                    if (FIRST[nt].find('') != FIRST[nt].end()) {
                        if (FIRST[nt].size() > 1) {
                            return false;
                        }
                        
                        if (FOLLOW[nt].find('') != FOLLOW[nt].end()) {
                            return false;
                        }
                    }
                    
                    for (char ch : FIRST[nt]) {
                        if (FIRST[prod1[0]].find(ch) != FIRST[prod1[0]].end() &&
                            FIRST[prod2[0]].find(ch) != FIRST[prod2[0]].end()) {
                            return false;
                        }
                    }
                }
            }
        }
    }
    
    return true;
}

int main() {
    if (is_LL1()) {
        cout << "The grammar is LL(1)." << endl;
    } else {
        cout << "The grammar is not LL(1)." << endl;
    }
    
    return 0;
}

File: dlp/python_dlp/prac5_LL1.py
def compute_FIRST(grammar):
    FIRST = {}
    changed = True

    while changed:
        changed = False
        for nt in grammar:
            FIRST[nt] = set()
            for prod in grammar[nt]:
                for symbol in prod:
                    if symbol in grammar:
                        prev_len = len(FIRST[nt])
                        FIRST[nt] |= FIRST.get(symbol, set())
                        if len(FIRST[nt]) != prev_len:
                            changed = True
                        if '' not in FIRST.get(symbol, set()):
                            break
                    else:
                        prev_len = len(FIRST[nt])
                        FIRST[nt].add(symbol)
                        if len(FIRST[nt]) != prev_len:
                            changed = True
                        break

    return FIRST

def compute_FOLLOW(grammar, FIRST):
    FOLLOW = {nt: set() for nt in grammar}
    FOLLOW[list(grammar.keys())[0]].add('$')

    changed = True
    while changed:
        changed = False
        for nt in grammar:
            for prod in grammar[nt]:
                prev_set = set()
                for symbol in reversed(prod):
                    if symbol in grammar:
                        prev_len = len(FOLLOW[symbol])
                        FOLLOW[symbol] |= FOLLOW[nt]
                        if len(FOLLOW[symbol]) != prev_len:
                            changed = True
                        if '' not in FIRST.get(symbol, set()):
                            break
                    else:
                        prev_len = len(prev_set)
                        prev_set.add(symbol)
                        if len(prev_set) != prev_len:
                            changed = True
                for symbol in reversed(prod):
                    if symbol in grammar:
                        if '' not in FIRST.get(symbol, set()):
                            break

    return FOLLOW

def is_LL1(grammar):
    FIRST = compute_FIRST(grammar)
    FOLLOW = compute_FOLLOW(grammar, FIRST)

    for nt in grammar:
        productions = grammar[nt]
        first_sets = [FIRST[nt] for prod in productions]
        for i in range(len(first_sets)):
            for j in range(i + 1, len(first_sets)):
                if (first_sets[i] & first_sets[j]):
                    return False

        for prod in productions:
            if '' in FIRST[nt]:
                if (FIRST[nt] & FOLLOW[nt]):
                    return False

    return True

# Define the grammar using a dictionary
grammar = {
    'E': ['aX'],
    'X': ['+aX', 'c']
}

if is_LL1(grammar):
    print("The grammar is LL(1).")
else:
    print("The grammar is not LL(1).")
